pkg/network/setup/configstate.go

```diff
@@ -45,40 +45,57 @@ func NewConfigState(cc cacheCreator, vmiUID string) ConfigState {
 // The configuration step is allowed to run only once. Any attempt to run it again will cause a critical error.
-func (c ConfigState) Run(podIfaceName string, discoverFunc func() error, configFunc func() error) error {
-	state, err := c.read(podIfaceName)
-	if err != nil {
-		return err
-	}
+func (c ConfigState) Run(nics []podNIC, discoverFunc func(*podNIC) error, configFunc func(*podNIC) error) error {
+	var pendingNICs []podNIC
+	for _, nic := range nics {
+		ifaceName := nic.podInterfaceName
+		state, err := c.read(ifaceName)
+		if err != nil {
+			return err
+		}
 
-	switch state {
-	case cache.PodIfaceNetworkPreparationStarted:
-		return neterrors.CreateCriticalNetworkError(
-			fmt.Errorf("pod interface %s network preparation cannot be restarted", podIfaceName),
-		)
-	case cache.PodIfaceNetworkPreparationFinished:
-		return nil
+		switch state {
+		case cache.PodIfaceNetworkPreparationPending:
+			pendingNICs = append(pendingNICs, nic)
+		case cache.PodIfaceNetworkPreparationStarted:
+			return neterrors.CreateCriticalNetworkError(
+				fmt.Errorf("pod interface %s network preparation cannot be restarted", ifaceName),
+			)
+		}
 	}
+	nics = pendingNICs
 
-	if ferr := discoverFunc(); ferr != nil {
-		return ferr
+	for i := range nics {
+		if ferr := discoverFunc(&nics[i]); ferr != nil {
+			return ferr
+		}
 	}
 
-	if werr := c.write(podIfaceName, cache.PodIfaceNetworkPreparationStarted); werr != nil {
-		return fmt.Errorf("failed to mark configuration as started for %s: %v", podIfaceName, werr)
+	for _, nic := range nics {
+		ifaceName := nic.podInterfaceName
+		if werr := c.write(ifaceName, cache.PodIfaceNetworkPreparationStarted); werr != nil {
+			return fmt.Errorf("failed to mark configuration as started for %s: %v", ifaceName, werr)
+		}
 	}
 
 	// The discovery step must be called *before* the configuration step, allowing it to persist/cache the
 	// original pod network status. The configuration step mutates the pod network.
-	if ferr := configFunc(); ferr != nil {
-		log.Log.Reason(err).Errorf("failed to configure pod network: %s", podIfaceName)
-		return neterrors.CreateCriticalNetworkError(ferr)
+	for _, nic := range nics {
+		nic := nic
+		ifaceName := nic.podInterfaceName
+		if ferr := configFunc(&nic); ferr != nil {
+			log.Log.Reason(ferr).Errorf("failed to configure pod network: %s", ifaceName)
+			return neterrors.CreateCriticalNetworkError(ferr)
+		}
 	}
 
-	if werr := c.write(podIfaceName, cache.PodIfaceNetworkPreparationFinished); werr != nil {
-		return neterrors.CreateCriticalNetworkError(
-			fmt.Errorf("failed to mark configuration as finished for %s: %v", podIfaceName, werr),
-		)
+	for _, nic := range nics {
+		ifaceName := nic.podInterfaceName
+		if werr := c.write(ifaceName, cache.PodIfaceNetworkPreparationFinished); werr != nil {
+			return neterrors.CreateCriticalNetworkError(
+				fmt.Errorf("failed to mark configuration as finished for %s: %v", ifaceName, werr),
+			)
+		}
```

Comment:
Wdyt about extracting each step loop to its own func ?

Reply:
Each loop body already does one thing, the rest is the error handling noise.
I think it will make it harder to follow if I try and extract it into more functions.

Reply:
I see, what about creating a loop over func ptrs that does the common error handling ?

Reply:
What will that save exactly? Calling the error factory?
We cannot really wrap error handling, with go we need to handle the error at each level anyway.

Reply:
Instead a lot of for loops, there will be just one loop on the list of the func ptrs
all of them return err anyhow

Reply:
I do not really understand what you suggest.
What I read now is accurate and it requires no special brain power to understand: Go over all the "nics", and do one thing. If we hide this, we will not understand that this is happening, which is not a good thing in this case IMO.

If you want to present your proposal as code, please do, maybe it will be convincing after-all.